File: src/Butil/Bit.Butil/Scripts/midi.ts
var BitButil = (window as any).BitButil = (window as any).BitButil || {};

(function (butil: any) {
    // One MIDIAccess per page: it is the permission grant, and re-requesting it with the same
    // options resolves to an equivalent object, so caching it keeps port ids stable.
    let _access: any = null;
    // The options the cached access was granted under - asking again with the same pair reuses it.
    let _accessSysex = false;
    let _accessSoftware = false;
    // MIDI port ids are strings the browser assigns and they stay valid for the life of the
    // access object, so - unlike the other bus APIs - .NET can address ports by their own id
    // and nothing has to be minted here.
    const _messageListeners: { [id: string]: { inputId: string | null, ports: any[], handler: EventListener } } = {};
    const _stateListeners: { [id: string]: EventListener } = {};

    // MIDI has no per-port 'added' event, so the access object's statechange is the only hook for
    // binding a subscription to a controller plugged in after it was made. One module-owned listener
    // covers every message subscription; it stays attached while any of them exists.
    let _reattachHandler: EventListener | null = null;

    butil.midi = {
        isSupported() { return typeof (navigator as any).requestMIDIAccess === 'function'; },
        requestAccess,
        getPorts,
        send,
        clear,
        subscribeMessages,
        unsubscribeMessages,
        subscribeStateChange,
        unsubscribeStateChange
    };

    function describePort(port: any) {
        return {
            id: port.id,
            name: port.name ?? null,
            manufacturer: port.manufacturer ?? null,
            version: port.version ?? null,
            type: port.type,
            state: port.state,
            connection: port.connection
        };
    }

    function list(map: any) {
        const out: any[] = [];
        // MIDIAccess.inputs/outputs are Maps; forEach is the only iteration ES2019 output can rely on.
        map?.forEach((port: any) => out.push(describePort(port)));
        return out;
    }

    function snapshot() {
        if (!_access) return null;
        return {
            sysexEnabled: !!_access.sysexEnabled,
            inputs: list(_access.inputs),
            outputs: list(_access.outputs)
        };
    }

    async function requestAccess(sysex: boolean, software: boolean) {
        if (typeof (navigator as any).requestMIDIAccess !== 'function') return null;

        // Re-requesting with the same options would resolve to an equivalent object, but a
        // different one - and every listener in _messageListeners/_stateListeners is attached to
        // the ports of the old one. Hand the cached access back instead.
        if (_access && _accessSysex === sysex && _accessSoftware === software) return snapshot();

        // Different options mean a different grant, so the old access object - and every port
        // listener attached through it - is replaced. The subscriptions themselves move across:
        // dropping them would leave .NET holding subscriptions that look live, never fire again, and
        // have nothing on either side able to say so.
        let access: any;
        try {
            access = await (navigator as any).requestMIDIAccess({ sysex, software });
        } catch {
            // A refused prompt - or a policy that blocks MIDI outright - rejects. That is "no access",
            // not a failure worth throwing across the boundary, and it must not disturb the access
            // already granted: every listener in _messageListeners is still attached to its ports.
            return null;
        }

        detachAll();
        _access = access;
        _accessSysex = sysex;
        _accessSoftware = software;
        attachAll();
        return snapshot();
    }

    // Detaches every listener from the outgoing access object without forgetting the subscriptions
    // behind them - attachAll() re-binds those to the incoming one.
    function detachAll() {
        if (!_access) return;

        stopWatchingPorts();
        for (const key of Object.keys(_messageListeners)) detachMessages(key);
        for (const key of Object.keys(_stateListeners)) _access.removeEventListener('statechange', _stateListeners[key]);
    }

    function attachAll() {
        if (!_access) return;

        for (const key of Object.keys(_stateListeners)) _access.addEventListener('statechange', _stateListeners[key]);
        if (Object.keys(_messageListeners).length) startWatchingPorts();
        for (const key of Object.keys(_messageListeners)) attachMessages(key);
    }

    function startWatchingPorts() {
        if (!_access || _reattachHandler) return;

        _reattachHandler = () => { for (const key of Object.keys(_messageListeners)) attachMessages(key); };
        _access.addEventListener('statechange', _reattachHandler);
    }

    function stopWatchingPorts() {
        if (!_reattachHandler) return;

        _access?.removeEventListener('statechange', _reattachHandler);
        _reattachHandler = null;
    }

    // The inputs a subscription covers right now: the one it named, or every input there is.
    function inputsFor(inputId: string | null) {
        if (!_access) return [];
        if (inputId) {
            const port = findPort(_access.inputs, inputId);
            return port ? [port] : [];
        }

        const ports: any[] = [];
        _access.inputs.forEach((port: any) => ports.push(port));
        return ports;
    }

    function detachMessages(subscriptionId: string) {
        const entry = _messageListeners[subscriptionId];
        if (!entry) return;

        for (const port of entry.ports) port.removeEventListener('midimessage', entry.handler);
        entry.ports = [];
    }

    // Binds a subscription to the inputs matching it at this moment. Safe to run again - the previous
    // bindings are dropped first, so a port that is still there ends up listened to exactly once.
    function attachMessages(subscriptionId: string) {
        const entry = _messageListeners[subscriptionId];
        if (!entry) return;

        detachMessages(subscriptionId);
        const ports = inputsFor(entry.inputId);
        for (const port of ports) {
            // The opens are independent, so they are started together rather than awaited one after
            // another - on a machine with a dozen virtual ports that is the difference between one
            // open latency and twelve. A midimessage cannot arrive before its port is open, so the
            // listener can go on straight away.
            try { if (port.connection !== 'open' && typeof port.open === 'function') void port.open()?.catch(() => { /* port went away */ }); }
            catch { /* port went away */ }
            port.addEventListener('midimessage', entry.handler);
        }
        entry.ports = ports;
    }

    // Null until requestAccess() has resolved - reading the port list is not itself a way to get
    // permission, and callers are expected to treat null as "ask first".
    function getPorts() { return snapshot(); }

    // MIDIAccess.inputs/outputs are maplike and keyed by exactly this id, so get() is the lookup -
    // there is no case where scanning the values would find a port that get() misses.
    function findPort(map: any, portId: string) {
        return map?.get(portId) ?? null;
    }

    async function send(outputId: string, data: Uint8Array, timestamp: number | null) {
        const output = findPort(_access?.outputs, outputId);
        if (!output) return false;
        try {
            // A closed port throws on send(); open() is idempotent and resolves immediately for
            // one that is already open.
            if (output.connection !== 'open' && typeof output.open === 'function') await output.open();
            // performance.now()-based timestamps only: 0/null means "as soon as possible".
            if (timestamp) output.send(data, timestamp);
            else output.send(data);
            return true;
        } catch { return false; }
    }

    // Drops every queued message that has not been sent yet - the panic button for a stuck note.
    function clear(outputId: string) {
        const output = findPort(_access?.outputs, outputId);
        if (!output || typeof output.clear !== 'function') return false;
        try { output.clear(); return true; } catch { return false; }
    }

    function subscribeMessages(subscriptionId: string, dotNetRef: any, inputId: string | null) {
        if (!_access) return false;
        // A named input has to exist. 'Every input' does not require one to be plugged in yet: the
        // subscription stands, and the statechange watch binds whatever turns up later.
        if (inputId && !findPort(_access.inputs, inputId)) return false;

        const handler: EventListener = (event: any) => {
            butil.utils.dispatch(dotNetRef, 'InvokeMidiMessage', subscriptionId, {
                portId: (event.target as any)?.id ?? '',
                data: event.data ? new Uint8Array(event.data) : new Uint8Array(0),
                timeStamp: event.timeStamp ?? 0
            });
        };

        // One entry per subscription regardless of how many ports it covers: the ports it is bound
        // to are recorded together, so unsubscribing detaches all of them and re-binding replaces
        // the whole set.
        _messageListeners[subscriptionId] = { inputId: inputId ?? null, ports: [], handler };
        startWatchingPorts();
        attachMessages(subscriptionId);
        return true;
    }

    function unsubscribeMessages(subscriptionId: string) {
        const entry = _messageListeners[subscriptionId];
        if (!entry) return;

        detachMessages(subscriptionId);
        delete _messageListeners[subscriptionId];
        if (Object.keys(_messageListeners).length === 0) stopWatchingPorts();
    }

    function subscribeStateChange(subscriptionId: string, dotNetRef: any) {
        if (!_access) return false;
        const handler: EventListener = (event: any) => {
            const port = event.port;
            butil.utils.dispatch(dotNetRef, 'InvokeMidiStateChange', subscriptionId, port ? describePort(port) : null);
        };
        _access.addEventListener('statechange', handler);
        _stateListeners[subscriptionId] = handler;
        return true;
    }

    function unsubscribeStateChange(subscriptionId: string) {
        const handler = _stateListeners[subscriptionId];
        if (!handler) return;
        delete _stateListeners[subscriptionId];
        _access?.removeEventListener('statechange', handler);
    }
}(BitButil));

File: src/Butil/Bit.Butil/Publics/Midi.cs
using System;
using System.Collections.Concurrent;
using System.Diagnostics.CodeAnalysis;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;
using Microsoft.JSInterop;

namespace Bit.Butil;

/// <summary>
/// Wraps the <see href="https://developer.mozilla.org/en-US/docs/Web/API/Web_MIDI_API">Web MIDI API</see>
/// (<c>navigator.requestMIDIAccess</c>): enumerate MIDI inputs and outputs, listen to the messages
/// a controller sends, and send messages to a synth.
/// </summary>
/// <remarks>
/// Chromium only, and only over HTTPS. <see cref="RequestAccess"/> prompts the user; asking for
/// system-exclusive access is a separate, stricter prompt, so leave it off unless the app really
/// does need to reprogram a device.
/// <br/>
/// Everything else on this service returns nothing useful until <see cref="RequestAccess"/> has
/// resolved - the port list is part of the grant, not something readable ahead of it.
/// </remarks>
[ButilService(typeof(Midi))]
public class Midi(IJSRuntime js) : IAsyncDisposable
{
    internal const string MessageMethodName = nameof(InvokeMidiMessage);
    internal const string StateChangeMethodName = nameof(InvokeMidiStateChange);

    private readonly ConcurrentDictionary<Guid, Action<MidiMessage>> _messageHandlers = new();
    private readonly ConcurrentDictionary<Guid, Action<MidiPortInfo?>> _stateHandlers = new();

    // Per-instance callback reference (see Keyboard): listeners are isolated per circuit / WASM app
    // and released on disposal - no static state, no cross-circuit leak.
    private DotNetObjectReference<Midi>? _dotNetRef;
    private DotNetObjectReference<Midi> DotNetRef => DotNetObjectReferenceHelper.GetOrCreate(ref _dotNetRef, this);

    /// <summary>True when the runtime exposes <c>navigator.requestMIDIAccess</c>.</summary>
    /// <remarks>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> IsSupported() => js.Invoke<bool>("BitButil.midi.isSupported");

    /// <summary>
    /// Asks the user for MIDI access and returns the ports it covers, or null when the prompt was
    /// refused. The resolved access is cached for the page, so calling this again with the same
    /// arguments is cheap and does not re-prompt.
    /// </summary>
    /// <remarks>
    /// Asking again with a different <paramref name="sysex"/> or <paramref name="software"/> is a
    /// different grant, so the browser replaces the access object and every port object under it.
    /// Subscriptions already handed out survive that: they are re-attached to the new ports, so a
    /// handler keeps firing and the <see cref="ButilSubscription"/> stays the way to detach it.
    /// A refusal leaves an access already granted untouched, subscriptions included.
    /// </remarks>
    /// <param name="sysex">
    /// Request system-exclusive messages as well. A separate and stricter permission - sysex can
    /// rewrite a device's firmware - so ask for it only when the app sends or reads sysex.
    /// </param>
    /// <param name="software">Include software synthesizers in the port list, not only hardware.</param>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(MidiAccessInfo))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(MidiPortInfo))]
    public ValueTask<MidiAccessInfo?> RequestAccess(bool sysex = false, bool software = false)
        => js.Invoke<MidiAccessInfo?>("BitButil.midi.requestAccess", sysex, software);

    /// <summary>
    /// The current port list. Null until <see cref="RequestAccess"/> has resolved - reading the
    /// ports is not itself a way to get permission.
    /// </summary>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(MidiAccessInfo))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(MidiPortInfo))]
    public ValueTask<MidiAccessInfo?> GetPorts() => js.Invoke<MidiAccessInfo?>("BitButil.midi.getPorts");

    /// <summary>
    /// Sends raw MIDI bytes to an output port, opening it first if needed.
    /// </summary>
    /// <param name="outputId">The <see cref="MidiPortInfo.Id"/> of the output.</param>
    /// <param name="data">The message bytes - status byte first.</param>
    /// <param name="timestamp">
    /// When to send it, on the <c>performance.now()</c> clock. Null or 0 sends immediately;
    /// scheduling ahead is how a sequencer gets timing the main thread cannot deliver.
    /// </param>
    public ValueTask<bool> Send(string outputId, byte[] data, double? timestamp = null)
        => js.Invoke<bool>("BitButil.midi.send", outputId, data, timestamp);

    /// <summary>
    /// Sends a note-on message on <paramref name="channel"/> (0-15). A velocity of 0 is a note-off
    /// by convention, and is what most devices actually send when a key is released.
    /// </summary>
    /// <exception cref="ArgumentOutOfRangeException"><paramref name="channel"/> is above 15.</exception>
    public ValueTask<bool> SendNoteOn(string outputId, byte note, byte velocity = 100, byte channel = 0)
    {
        ThrowIfNotAChannel(channel);
        return Send(outputId, [(byte)(0x90 | channel), note, velocity]);
    }

    /// <summary>Sends a note-off message on <paramref name="channel"/> (0-15).</summary>
    /// <exception cref="ArgumentOutOfRangeException"><paramref name="channel"/> is above 15.</exception>
    public ValueTask<bool> SendNoteOff(string outputId, byte note, byte velocity = 0, byte channel = 0)
    {
        ThrowIfNotAChannel(channel);
        return Send(outputId, [(byte)(0x80 | channel), note, velocity]);
    }

    // Masking a bad channel down to 0-15 would send the message on some other channel and look like
    // the device ignored it; the caller's arithmetic is what is wrong, so it is raised as such.
    private static void ThrowIfNotAChannel(byte channel, [CallerArgumentExpression(nameof(channel))] string? paramName = null)
    {
        if (channel > 15) throw new ArgumentOutOfRangeException(paramName, channel, "A MIDI channel is 0-15.");
    }

    /// <summary>
    /// Drops every message queued for the port that has not been sent yet - the way out of a note
    /// left hanging by a cancelled sequence.
    /// </summary>
    public ValueTask<bool> Clear(string outputId) => js.Invoke<bool>("BitButil.midi.clear", outputId);

    /// <summary>
    /// Invoked from JS for each incoming MIDI message. Public + <see cref="JSInvokableAttribute"/>
    /// so it can be dispatched through the per-instance <see cref="DotNetObjectReference{T}"/>.
    /// </summary>
    [JSInvokable(MessageMethodName)]
    public void InvokeMidiMessage(Guid id, MidiMessage message)
    {
        if (_messageHandlers.TryGetValue(id, out var handler)) handler.Invoke(message);
    }

    /// <summary>Invoked from JS on a port's <c>statechange</c>. See <see cref="InvokeMidiMessage"/>.</summary>
    [JSInvokable(StateChangeMethodName)]
    public void InvokeMidiStateChange(Guid id, MidiPortInfo? port)
    {
        if (_stateHandlers.TryGetValue(id, out var handler)) handler.Invoke(port);
    }

    /// <summary>
    /// Listens to the messages arriving on one input, or on every input at once.
    /// </summary>
    /// <param name="handler">Called for each message, with the port it came from.</param>
    /// <param name="inputId">
    /// The input to listen to, or null for all of them - which is usually what an app wants, since
    /// the user's controller is whichever one they touch.
    /// </param>
    /// <returns>A subscription - dispose it to detach the listener.</returns>
    /// <remarks>
    /// Listening to all of them is a standing subscription, not a snapshot: a controller plugged in
    /// afterwards is picked up too, and subscribing before anything is plugged in is allowed. A
    /// named <paramref name="inputId"/>, by contrast, has to exist at the time of the call.
    /// </remarks>
    /// <exception cref="InvalidOperationException">The listener was not attached - no MIDI access, or no such input.</exception>
    [DynamicDependency(nameof(InvokeMidiMessage), typeof(Midi))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(MidiMessage))]
    public async ValueTask<ButilSubscription> SubscribeMessages(Action<MidiMessage> handler, string? inputId = null)
    {
        return await ButilSubscriptionHelper.Register(_messageHandlers, handler,
                                                      id => js.InvokeRegister("BitButil.midi.subscribeMessages", id, DotNetRef, inputId),
                                                      id => js.InvokeVoid("BitButil.midi.unsubscribeMessages", id),
                                                      "The MIDI message listener could not be attached - call RequestAccess() first, and make sure the named input exists.");
    }

    /// <summary>
    /// Watches ports being connected and disconnected. The handler gets the port that changed, so
    /// re-reading the whole list is only necessary when the app tracks more than one.
    /// </summary>
    /// <returns>A subscription - dispose it to detach the listener.</returns>
    /// <exception cref="InvalidOperationException">The listener was not attached - no MIDI access.</exception>
    [DynamicDependency(nameof(InvokeMidiStateChange), typeof(Midi))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(MidiPortInfo))]
    public async ValueTask<ButilSubscription> SubscribeStateChange(Action<MidiPortInfo?> handler)
    {
        return await ButilSubscriptionHelper.Register(_stateHandlers, handler,
                                                      id => js.InvokeRegister("BitButil.midi.subscribeStateChange", id, DotNetRef),
                                                      id => js.InvokeVoid("BitButil.midi.unsubscribeStateChange", id),
                                                      "The MIDI state-change listener could not be attached - call RequestAccess() first.");
    }

    /// <summary>Detaches every listener this instance attached and releases its interop reference.</summary>
    public async ValueTask DisposeAsync()
    {
        try
        {
            foreach (var id in _messageHandlers.Keys.ToArray())
            {
                _messageHandlers.TryRemove(id, out _);
                await js.InvokeVoid("BitButil.midi.unsubscribeMessages", id);
            }

            foreach (var id in _stateHandlers.Keys.ToArray())
            {
                _stateHandlers.TryRemove(id, out _);
                await js.InvokeVoid("BitButil.midi.unsubscribeStateChange", id);
            }
        }
        catch (Exception ex) when (ex.IsIgnorableDisposalException()) { } // teardown: circuit gone, cancelled, or already disposed
        finally
        {
            _dotNetRef?.Dispose();
            _dotNetRef = null;
        }

        GC.SuppressFinalize(this);
    }
}
